timesync/ticker.go

```diff
@@ -179,22 +182,28 @@ func init() {
 	close(closedChan)
 }
 
-// AwaitLayer returns a channel that will be signaled when layer id layerID was ticked by the clock, or if this layer has passed
-// while sleeping. it does so by closing the returned channel.
-func (t *Ticker) AwaitLayer(layerID types.LayerID) chan struct{} {
+// AwaitLayer returns a context that will be canceled when layer with given id is ticked by the clock,
+// or if this layer has passed while sleeping.
+func (t *Ticker) AwaitLayer(ctx context.Context, layerID types.LayerID) context.Context {
```

Comment:
I think this method could just be

```golang
func (t *Ticker) AwaitLayer(ctx context.Context, layerID types.LayerID) context.Context {
	t.mu.Lock()
	defer t.mu.Unlock()

	layerCtx, cancel := context.WithDeadline(t.LayerToTime(layerID))
        t.layersSubscriptions[layerID] = append(t.layersSubscriptions[layerID], cancel)
	return layerCtx
}

Reply:
On second thought, this could maybe break things.

At the moment `Notify` needs to be called explicitly for the listeners to be made aware that a layer has been reached. `Notify` can be called late so listeners can be notified late, this might be required.

`Notify` errors if it is called too late, but still signals to all listeners that their layer has been reached. This also means that the existing code and the changed code in this PR have a sort of race condition:

- Assume `LayerToTime` for layer `L` returns time `T`
- All go routines calling `AwaitLayer` before T receive a `context` that is not cancelled (or a channel that is not closed)
- All go routines calling `AwaitLayer` after T receive a `context` that is cancelled (or a channel that is closed)
- The early callers receive their signal only after `Notify` is called, which is expected to be up to `sendTickThreshold` late; it can be even later then that, but then `Notify` errors

Reply:
In my opinion there are 2 possible ways to fix this issue:

- **simple & small change**: `AwaitLayer` doesn't check `t.LayerToTime` to determine if the returned `context` should already be cancelled or not, but rather `t.lastTickedLayer`.
- **probably bigger change**: Remove the canceling of contexts in `Notify` and use an implementation for `AwaitLayer` similar to my suggestion above. `Notify` then only notifies the subscribers who called `Subscribe`, while listeners of `AwaitLayer` handle themselves.

The relevant code part in `Notify` that imo can be removed with my suggested change is the following:

```golang
	for l := t.lastTickedLayer; !l.After(layer); l = l.Add(1) {
		if cancels, found := t.layersSubscriptions[l]; found {
			for _, cancel := range cancels {
				cancel()
			}
			delete(t.layersSubscriptions, l)
		}
	}
```

Reply:
@countvonzero, @dshulyak What do you think?

Reply:
@poszu @fasmat 
can either of you elaborate in the PR description what was wrong with the original channel approach? thanks

Reply:
still don't understand. 

`AwaitLayer` is used in many places. the pattern in the code code either waits for a layer to arrive (in which case the channel unblocks) or waits for the app to shutdown (`context.Done()` unblocks). what is the advantage combining the two?

the disadvantage for me is that i wouldn't know whether it's the layer time arrives or the app shutdown. the former i need to do something, the latter i need to stop.

Reply:
> So that AwaitLayer stops waiting for a new layer when the app shuts down.

where does it happen? can you point me to the code where AwaitLayer is unblocked if the app shuts down?

Reply:
It's not unblocked when the app shuts down. Still, both events can happen at the roughly same time. Then `select` will choose randomly.